python/ray/air/util/tensor_extensions/utils.py

```diff
@@ -42,3 +44,34 @@ def _create_strict_ragged_ndarray(values: Any) -> np.ndarray:
     # Try to fill the 1D array of pointers with the (ragged) tensors.
     arr[:] = list(values)
     return arr
+
+
+def _create_possibly_ragged_ndarray(
+    values: Union[np.ndarray, ABCSeries, Sequence[Any]]
+) -> np.ndarray:
+    """
+    Create a possibly ragged ndarray.
+
+    Using the np.array() constructor will fail to construct a ragged ndarray that has a
+    uniform first dimension (e.g. uniform channel dimension in imagery). This function
+    catches this failure and tries a create-and-fill method to construct the ragged
+    ndarray.
+    """
+    try:
+        with warnings.catch_warnings():
+            # For NumPy < 1.24, constructing a ragged ndarray directly via
+            # `np.array(...)` without the `dtype=object` parameter will raise a
+            # VisibleDeprecationWarning which we suppress.
+            # More details: https://stackoverflow.com/q/63097829
+            warnings.simplefilter("ignore", category=np.VisibleDeprecationWarning)
+            return np.array(values, copy=False)
+    except ValueError as e:
+        # For NumPy >= 1.24, constructing a ragged ndarray directly via `np.array(...)`
+        # without the `dtype=object` parameter will raise a ValueError.
+        # https://github.com/numpy/numpy/pull/22004
+        if "The requested array has an inhomogeneous shape" in str(e):
```

Comment:
OOC why'd we change the string from `"could not broadcast input array from shape"` to `"The requested array has an inhomogeneous shape"`?

Reply:
It seems like NumPy has changed the error message string, since the full message that I get when I run [this reproducible script](https://github.com/ray-project/ray/issues/31573#issuecomment-1377843926) on NumPy 1.24.0 is:
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/sjl/src/ray/python/ray/air/util/tensor_extensions/utils.py", line 77, in _create_possibly_ragged_ndarray
    raise e from None
  File "/Users/sjl/src/ray/python/ray/air/util/tensor_extensions/utils.py", line 67, in _create_possibly_ragged_ndarray
    return np.array(values, copy=False)
ValueError: setting an array element with a sequence. The requested array has an inhomogeneous shape after 1 dimensions. The detected shape was (3,) + inhomogeneous part.
```

Reply:
If I use an earlier version of NumPy, would `_create_possibly_ragged_ndarray ` raise an error instead of creating a ragged array?

Reply:
For earlier versions of NumPy, this would normally emit the `VisibleDeprecationWarning` warning, but we now catch this with the changes in this PR.

Reply:
Oh I see what you maen @bveeramani . When using an earlier NumPy version, it will still output the old error message `"could not broadcast input array from shape"`, so I will need to keep both versions of the error strings to check.

Reply:
> The actual ragged array construction still occurs, with the default np.array(...) constructor (it doesn't go into the except case since a ValueError is not thrown) 

I'm confused. Wouldn't a `ValueError` still be thrown if there's a broadcast error? And how would it go through the `_create_strict_ragged_ndarray` code path if the error message changed?

Like, if I'm using NumPy 1.23 and call
```
>>> _create_possibly_ragged_tensor([np.zeros([3, 32, 32]), np.zeros([3, 64, 64])])
```

Then `_create_possibly_ragged_tensor` would call

```
np.array([np.zeros([3, 32, 32]), np.zeros([3, 64, 64])], copy=False)
```

And then `numpy` would raise:

```
ValueError: could not broadcast input array from shape (3,32,32) into shape (3,)
```

Since `"The requested array has an inhomogeneous shape" ` isn't in this error message, we would re-raise the erorr with `raise e from None` instead of falling back to `_create_strict_ragged_ndarray`.

Reply:
Oh, just saw your message. Okay, gotcha. Sounds good

Reply:
yeah thanks for the catch! Sorry I misunderstood this earlier